File: Machine Learning/1.1.5 Performance evaluation/ML.R
library("ROCR")

#读取数据
data_mat <- read.csv("/Users/xinzhe/Desktop/Lulab/毕设/gitbook/BreastCancer.csv", header = T)
for(i in 1:length(data_mat[,7])){
  if(is.na(data_mat[i,7]))
    data_mat[i,7] <- 0
}
rm(i)

#数据归一化
X <- rep(0,9*699)
dim(X) <- c(699, 9)
for(i in 2:(dim(data_mat)[2]-1)){
  X[,i-1] <- (data_mat[, i] - min(data_mat[, i])) / (max(data_mat[, i])-min(data_mat[, i]))
}
rm(i)

#数据切分
label_mat <- as.character(data_mat[, 11])
X <- data.frame(X, label_mat)
set.seed(520)
train <- sample(nrow(X), 0.75*nrow(X))
X_train <- X[train,]
Y_train <- as.character(X_train[, 10])
X_test <- X[-train,]
Y_test <- as.character(X_test[, 10])

#模型训练测试
logistic <- glm(label_mat ~ ., data = X_train, family = binomial())

#预测
predicted <- as.numeric(predict(logistic, as.data.frame(X_test), type = 'response') > 0.5)
pred <- prediction(predicted, Y_test)
auc <- performance(pred, 'auc')@y.values

# 输出AUC
auc

#画ROC曲线
roc <- performance(pred, "tpr", "fpr")
plot(roc, main = 'ROC Curve', col = "blue", lwd = 2)
abline(a = 0, b = 1, lwd = 2, col="grey", lty = 2)
legend("bottomright", inset=0.01,c("ROC", "Random Chance"), col=c("blue", "grey"), lty = c(1, 2))

File: Machine Learning/1.1.5 Performance evaluation/ML.py
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import roc_curve, auc
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

#读取数据
data_mat = []
label_mat = []
with open('/Users/xinzhe/Desktop/Lulab/毕设/gitbook/BreastCancer.csv', 'r') as f:
    for line in f.readlines()[1:]:
        line_arr = line.strip().split(',')
        if line_arr[6] == 'NA':
            line_arr[6] = 0
        data_mat.append([float(line_arr[1]), float(line_arr[2]), float(line_arr[3]), float(line_arr[4]),
        float(line_arr[5]), float(line_arr[6]), float(line_arr[7]), float(line_arr[8]), float(line_arr[9])])
        label_mat.append(line_arr[10])

#数据归一化
scaler = MinMaxScaler()
X, Y = scaler.fit_transform(data_mat), label_mat

#数据切分
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, random_state = 520, test_size = 0.25)

#模型训练测试
log_reg = LogisticRegression()
log_reg.fit(X_train, Y_train)

#预测
Y_predict = log_reg.predict(X_test)
proba = log_reg.predict_proba(X_test).T[1]
fpr, tpr, thresholds = roc_curve(Y_test, proba, pos_label = "malignant")
roc_auc = auc(fpr, tpr)

# 输出AUC
print(roc_auc)

#画ROC曲线
plt.plot(fpr, tpr, '-', color='b', label='ROC', lw=2)
plt.plot([0, 1], [0, 1], '--', color=(0.6, 0.6, 0.6), label='Random Chance')
plt.xlim([0, 1])
plt.ylim([0, 1])
plt.title('ROC curve')
plt.xlabel('FPR')
plt.ylabel('TPR')
plt.legend()
plt.show()
plt.close()
